torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,510 @@
+        return mem
+
+    def get_mem_consumed(self, st: torch.UntypedStorage) -> int:
+        if st.size() != self.size:
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+    @staticmethod
+    def get_untyped_storage(t: torch.Tensor) -> torch.UntypedStorage:
+        if isinstance(t, DTensor):
+            t = t._local_tensor
+        return t.untyped_storage()
+
+    @classmethod
+    def get_winfo_if_exists(
+        cls, st: torch.UntypedStorage, WINFO: WeakIdKeyDictionary
+    ) -> Optional[Self]:
+        winfo = WINFO.get(st, None)
+        if winfo is None:
+            return None
+        return winfo
+
+    @classmethod
+    def create_winfo(
+        cls, st: torch.UntypedStorage, reftype: RefType, WINFO: WeakIdKeyDictionary
+    ) -> Self:
+        winfo = cls(st.size(), st.element_size(), st.device, reftype)
+        WINFO[st] = winfo
+        return winfo
+
+    @classmethod
+    def update_and_maybe_create_winfo(
+        cls,
+        t: torch.Tensor,
+        reftype: RefType,
+        WINFO: WeakIdKeyDictionary,
+        update_existing: bool = False,
+    ) -> Self:
+        st = cls.get_untyped_storage(t)
+        if winfo := cls.get_winfo_if_exists(st, WINFO):
+            winfo.reftype = reftype
+            return winfo
+        elif update_existing:
+            raise ValueError("No existing winfo found")
+        else:
+            winfo = cls.create_winfo(st, reftype, WINFO)
+            return winfo
+
+
+def _get_mem_divisor(units: str) -> int:
+    return {"B": 1, "KB": 2**10, "MB": 2**20, "GB": 2**30}.get(units, 1)
+
+
+def _rounding_fn(value: int, divisor: int, precision: int) -> Union[float, int]:
+    return value if divisor == 1 else round(value / divisor, precision)
+
+
+def _print_snapshot(snapshot: Dict[torch.device, Dict[str, int]], units: str):
+    divisor = _get_mem_divisor(units)
+    for dev, dev_snap in snapshot.items():
+        print(
+            f"Device: {dev}",
+            *(
+                f"\t{k}: {_rounding_fn(v, divisor, 2)} {units}"
+                for k, v in dev_snap.items()
+            ),
+            sep="\n",
+        )
+
+
+class MemTracker(TorchDispatchMode):
+    """
+    A TorchDispatchMode that tracks the memory usage of parameters, buffers, activations, gradients, and optimizer states
+    within its context. It captures "memory snapshots" of the module's memory usage at various points in time
+    [pre_fw, post_fw, pre_bw, post_bw, peak_fw, peak_bw].
+
+    Attributes:
+        memory_tracking: A weakref key dictionary to store the memory statistics of each module. Each key
+        is a weak reference to a module, and each value is a `ModMemStats` object that stores the memory
+        statistics of the module.
+
+    Note:
+        The MemTracker should be used as a context manager. The modules, optimizers, and any other tensors created within
```

Comment:
nit: I would rephrase this as Tensor focused. And only say that you can register external Tensor or stateful objects such as module/opt/etc